dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/WorkContextLoader.java

```diff
@@ -0,0 +1,177 @@
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.hibernate.SessionFactory;
+import org.hisp.dhis.dxf2.common.ImportOptions;
+import org.hisp.dhis.dxf2.events.event.Event;
+import org.hisp.dhis.hibernate.HibernateUtils;
+import org.hisp.dhis.program.ProgramStageInstance;
+import org.hisp.dhis.trackedentity.TrackedEntityInstance;
+import org.hisp.dhis.user.User;
+import org.hisp.dhis.user.UserCredentials;
+import org.springframework.stereotype.Component;
+import org.springframework.transaction.annotation.Transactional;
+
+/**
+ * @author Luciano Fiandesio
+ */
+@Component
+public class WorkContextLoader
+{
+    private final ProgramSupplier programSupplier;
+
+    private final OrganisationUnitSupplier organisationUnitSupplier;
+
+    private final TrackedEntityInstanceSupplier trackedEntityInstanceSupplier;
+
+    private final ProgramInstanceSupplier programInstanceSupplier;
+
+    private final ProgramStageInstanceSupplier programStageInstanceSupplier;
+
+    private final CategoryOptionComboSupplier categoryOptionComboSupplier;
+
+    private final DataElementSupplier dataElementSupplier;
+
+    private final NoteSupplier noteSupplier;
+
+    private final AssignedUserSupplier assignedUserSupplier;
+
+    private final ServiceDelegatorSupplier serviceDelegatorSupplier;
+
+    private final static UidGenerator uidGen = new UidGenerator();
+
+    private final SessionFactory sessionFactory;
+
+    public WorkContextLoader(
+    // @formatter:off
+        ProgramSupplier programSupplier,
+        OrganisationUnitSupplier organisationUnitSupplier,
+        TrackedEntityInstanceSupplier trackedEntityInstanceSupplier,
+        ProgramInstanceSupplier programInstanceSupplier,
+        ProgramStageInstanceSupplier programStageInstanceSupplier,
+        CategoryOptionComboSupplier categoryOptionComboSupplier,
+        DataElementSupplier dataElementSupplier,
+        NoteSupplier noteSupplier,
+        AssignedUserSupplier assignedUserSupplier,
+        ServiceDelegatorSupplier serviceDelegatorSupplier,
+        SessionFactory sessionFactory
+    // @formatter:on
+    )
+    {
+        this.programSupplier = programSupplier;
+        this.organisationUnitSupplier = organisationUnitSupplier;
+        this.trackedEntityInstanceSupplier = trackedEntityInstanceSupplier;
+        this.programInstanceSupplier = programInstanceSupplier;
+        this.programStageInstanceSupplier = programStageInstanceSupplier;
+        this.categoryOptionComboSupplier = categoryOptionComboSupplier;
+        this.dataElementSupplier = dataElementSupplier;
+        this.noteSupplier = noteSupplier;
+        this.assignedUserSupplier = assignedUserSupplier;
+        this.serviceDelegatorSupplier = serviceDelegatorSupplier;
+        this.sessionFactory = sessionFactory;
+    }
+
+    @Transactional( readOnly = true )
+    public WorkContext load( ImportOptions importOptions, List<Event> events )
+    {
+        sessionFactory.getCurrentSession().flush();
```

Comment:
Since, this code could be called during an import of Tei -> Enrollment, we need to be sure that these entities are flushed before loading the data in cache.